File: examples/cpp/private/example_current.cpp
#include <iostream>
#include <csignal>
#include <atomic>

std::atomic<bool> keep_running{true};

void sigint_handler(int) {
    keep_running = false;
}
#include <thread>
#include <vector>
#include <chrono>
#include <cmath>
#include <iomanip>
#include "robstride/robstride_motor.h"
#include "robstride/robstride_transport.h"

using namespace robstride;

bool setup_motor(Motor& motor, const std::string& name) {
    std::cout << "[" << name << "] Pinging motor..." << std::endl;
    if (motor.ping() != Result::SUCCESS) {
        std::cerr << "[" << name << "] Failed to ping motor. Is it connected and powered?" << std::endl;
        return false;
    }
    
    auto status = motor.get_status();
    std::cout << "[" << name << "] Ping successful! MCU ID: 0x" << std::hex << std::uppercase << status.mcu_id << std::dec << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "[" << name << "] Clearing errors..." << std::endl;
    if (motor.clear_error() != Result::SUCCESS) {
        std::cerr << "[" << name << "] Failed to clear errors." << std::endl;
        return false;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "[" << name << "] Setting current position as absolute zero..." << std::endl;
    if (motor.set_zero() != Result::SUCCESS) {
        std::cerr << "[" << name << "] Failed to set zero position." << std::endl;
        return false;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "[" << name << "] Setting up CURRENT (Direct Current) mode..." << std::endl;
    if (motor.setup_mode(RunMode::CURRENT) != Result::SUCCESS) {
        std::cerr << "[" << name << "] Failed to enter CURRENT mode." << std::endl;
        return false;
    }

    std::cout << "[" << name << "] Enabling Motor..." << std::endl;
    if (motor.enable() != Result::SUCCESS) {
        std::cerr << "[" << name << "] Failed to enable motor." << std::endl;
        return false;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    return true;
}

int main() {
    std::signal(SIGINT, sigint_handler);
    std::cout << "Connecting to CAN interface..." << std::endl;
    auto transport = std::make_shared<SocketCANTransport>("can0");
    if (transport->connect() != Result::SUCCESS) {
        std::cerr << "Failed to connect to can0. Is it up?" << std::endl;
        return 1;
    }
    
    // 1. Initialize Two Motors on the shared transport
    Motor motor5(5, 0, Protocol::PRIVATE, PROFILE_RS00, transport);
    Motor motor6(6, 0, Protocol::PRIVATE, PROFILE_RS00, transport);
    std::vector<Motor*> motors = {&motor5, &motor6};

    // 2. Start shared receive thread
    std::thread rx_thread([transport, motors]() {
        Motor::receive_all(transport, motors);
    });

    // 3. Setup both motors
    if (!setup_motor(motor5, "Motor 5") || !setup_motor(motor6, "Motor 6")) {
        transport->disconnect();
        if (rx_thread.joinable()) rx_thread.join();
        return 1;
    }

    std::cout << "Starting High-Frequency Direct Current Control. Press Ctrl+C to stop." << std::endl;
    
    auto start_time = std::chrono::steady_clock::now();
    
    float last_torque5 = -999.0f;
    float last_torque6 = -999.0f;

    while (transport->is_connected() && keep_running) {
        auto now = std::chrono::steady_clock::now();
        double t = std::chrono::duration<double>(now - start_time).count();
        
        // Generate a smooth 0.5Hz sine wave with amplitude of 2.0 Amps
        float target_iq = static_cast<float>(2.0 * std::sin(2.0 * M_PI * 0.5 * t));
        
        auto res5 = motor5.send_current_control(target_iq);
        if (res5 != robstride::Result::SUCCESS) {
            std::cerr << "\n\n[FATAL] Transmission failed on Motor 5 with error code: " << static_cast<int>(res5) << "! Aborting loop." << std::endl;
            break;
        }
        auto res6 = motor6.send_current_control(target_iq);
        if (res6 != robstride::Result::SUCCESS) {
            std::cerr << "\n\n[FATAL] Transmission failed on Motor 6 with error code: " << static_cast<int>(res6) << "! Aborting loop." << std::endl;
            break;
        }
        
        auto s5 = motor5.get_status();
        auto s6 = motor6.get_status();
        
        if (s5.torque != last_torque5 || s6.torque != last_torque6) {
            std::cout << "\r" << std::fixed << std::setprecision(2)
                      << "M5 TgtCur:" << std::setw(6) << target_iq << "A"
                      << " ActTrq:" << std::setw(5) << s5.torque << "Nm"
                      << " Pos:" << std::setw(5) << s5.angle 
                      << " Vel:" << std::setw(5) << s5.velocity << " | "
                      << "M6 TgtCur:" << std::setw(6) << target_iq << "A"
                      << " ActTrq:" << std::setw(5) << s6.torque << "Nm"
                      << " Pos:" << std::setw(5) << s6.angle 
                      << " Vel:" << std::setw(5) << s6.velocity 
                      << std::flush;
                      
            last_torque5 = s5.torque;
            last_torque6 = s6.torque;
        }
        
        // 1000Hz loop
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }

    std::cout << "\nStopping Motors..." << std::endl;
    motor5.stop();
    motor6.stop();
    transport->disconnect();
    
    if (rx_thread.joinable()) {
        rx_thread.join();
    }
    
    std::cout << "Done." << std::endl;
    return 0;
}

File: examples/python/private/example_current.py
#!/usr/bin/env python3
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../build')))
import pyrobstride
import time
import threading
import math

def rx_thread_func(transport, motors):
    pyrobstride.Motor.receive_all(transport, motors)

def setup_motor(motor, name):
    print(f"[{name}] Pinging motor...")
    if motor.ping() != pyrobstride.Result.SUCCESS:
        print(f"[{name}] Failed to ping motor. Is it connected and powered?")
        return False
    
    status = motor.get_status()
    print(f"[{name}] Ping successful! MCU ID: 0x{status.mcu_id:X}")
    time.sleep(0.1)

    print(f"[{name}] Clearing errors...")
    if motor.clear_error() != pyrobstride.Result.SUCCESS:
        return False
    time.sleep(0.1)

    print(f"[{name}] Setting current position as absolute zero...")
    if motor.set_zero() != pyrobstride.Result.SUCCESS:
        return False
    time.sleep(0.1)

    print(f"[{name}] Setting up CURRENT (Direct Current) mode...")
    if motor.setup_mode(pyrobstride.RunMode.CURRENT) != pyrobstride.Result.SUCCESS:
        print(f"[{name}] Failed to enter CURRENT mode.")
        return False

    print(f"[{name}] Enabling pyrobstride.Motor...")
    if motor.enable() != pyrobstride.Result.SUCCESS:
        return False
    time.sleep(0.5)
    return True

def main():
    print("Connecting to CAN interface...")
    transport = pyrobstride.SocketCANTransport("can0")
    if transport.connect() != pyrobstride.Result.SUCCESS:
        print("Failed to connect to can0. Is it up?")
        return

    motor5 = pyrobstride.Motor(5, 0, pyrobstride.Protocol.PRIVATE, pyrobstride.PROFILE_RS00, transport)
    motor6 = pyrobstride.Motor(6, 0, pyrobstride.Protocol.PRIVATE, pyrobstride.PROFILE_RS00, transport)
    motors = [motor5, motor6]

    rx_thread = threading.Thread(target=rx_thread_func, args=(transport, motors), daemon=True)
    rx_thread.start()

    if not setup_motor(motor5, "Motor 5") or not setup_motor(motor6, "Motor 6"):
        transport.disconnect()
        return

    print("Starting High-Frequency Direct Current Control. Press Ctrl+C to stop.")
    
    start_time = time.time()
    last_torque5 = 0.0
    last_torque6 = 0.0
    
    try:
        while True:
            t = time.time() - start_time
            # Generate a smooth 0.5Hz sine wave with amplitude of 2.0 Amps
            target_iq = 2.0 * math.sin(2.0 * math.pi * 0.5 * t)
            
            res5 = motor5.send_current_control(target_iq)
            if res5 != pyrobstride.Result.SUCCESS:
                print(f"\n[FATAL] Transmission failed on Motor 5 with error: {res5.name}! Aborting loop.")
                break
                
            res6 = motor6.send_current_control(target_iq)
            if res6 != pyrobstride.Result.SUCCESS:
                print(f"\n[FATAL] Transmission failed on Motor 6 with error: {res6.name}! Aborting loop.")
                break
                
            status5 = motor5.get_status()
            status6 = motor6.get_status()
            if status5.torque != last_torque5 or status6.torque != last_torque6:
                out = (f"\rM5 TgtCur:{target_iq:>6.2f}A ActTrq:{status5.torque:>5.2f}Nm Pos:{status5.angle:>5.2f} Vel:{status5.velocity:>5.2f} | "
                       f"M6 TgtCur:{target_iq:>6.2f}A ActTrq:{status6.torque:>5.2f}Nm Pos:{status6.angle:>5.2f} Vel:{status6.velocity:>5.2f}")
                print(out, end="", flush=True)
                last_torque5 = status5.torque
                last_torque6 = status6.torque
            
            # Yield slightly targeting ~1000Hz loop
            time.sleep(0.001)

    except KeyboardInterrupt:
        print("\nStopping loop due to Ctrl+C...")

    print("Stopping Motors...")
    motor5.stop()
    motor6.stop()
    transport.disconnect()

if __name__ == '__main__':
    main()
